feat(core): allow defining multiple named configurations in one config file (#6201)
mikro-orm/mikro-orm: docs/docs/quick-start.md

$ yarn mikro-orm
```

> For CLI to be able to access your database, you will need to create a configuration file that exports your ORM configuration(s).

Reviewer: yeah we can leave that for a separate PR

By default, the following paths, relative to the current working directory, are searched in this order:
```
./src/mikro-orm.config.ts
./mikro-orm.config.ts
./dist/mikro-orm.config.js
./build/mikro-orm.config.js
./src/mikro-orm.config.js

> ./mikro-orm.config.js

Reviewer: yeah agreed on the order. but cant say i would like to have this extensive list in the docs :]

- `MIKRO_ORM_CLI_ALWAYS_ALLOW_TS`: enable `.ts` files to use without ts-node
- `MIKRO_ORM_CLI_VERBOSE`: enable verbose logging (e.g. print queries used in seeder or schema diffing)

> MikroORM will always try to load the first available config file, based on the order in `configPaths`. When you have `useTsNode` disabled or `ts-node` is not already registered nor detected, TS config files will be ignored.

Reviewer: ```suggestion
MikroORM will always try to load the first available config file, based on the order in `configPaths`. When you have `useTsNode` explicitly disabled or `ts-node` is not already registered nor detected, TS config files will be ignored.
```

You can also specify the config path via `--config` option:

```sh
$ npx mikro-orm debug --config ./my-config.ts
```

Since v6.3, the CLI will always try to use TS config file, even without explicitly enabling it via `useTsNode` flag in your `package.json` file. You can still use it to disable the TS support explicitly. Keep in mind that having `ts-node` installed is still required for the TS support to work. The `useTsNode` has effect only on the CLI.

> Your configuration file may export multiple configuration objects in an array. The different configurations must heve a "name" in them. If no name is specified, it is treated as the name "default". You can use the `MIKRO_ORM_CONFIG_NAME` environment variable or the `--configName` command line option to pick a configuration with a particular name to use for the CLI. See [below](#configuration-file-structure) for details on the config object.

Reviewer: ```suggestion
Your configuration file may export multiple configuration objects in an array. The different configurations must have a `contextName` option. If no context name is specified, it is treated as the name "default". You can use the `MIKRO_ORM_CONFIG_NAME` environment variable or the `--configName` command line option to pick a configuration with a particular name to use for the CLI. See [below](#configuration-file-structure) for details on the config object.
```

i also wonder if we should change/unify the `configName` and corresponding env var to use the existing `contextName` instead?

                                    diff
  mikro-orm migration:up            Migrate up to the latest version
  mikro-orm migration:down          Migrate one step down
  mikro-orm migration:list          List all executed migrations
  mikro-orm migration:check         Check if migrations are needed. Useful for
                                    bash scripts.
  mikro-orm migration:pending       List all pending migrations
  mikro-orm migration:fresh         Clear the database and rerun all migrations
  mikro-orm debug                   Debug CLI configuration

Options:
      --config      Set path to the ORM configuration file               [array]
      --configName  Set name of config to load out of the ORM configuration
                    file. Used when config file exports an array or a function
                                                   [string] [default: "default"]
  -v, --version     Show version number                                [boolean]
  -h, --help        Show help                                          [boolean]
  -v, --version  Show version number                                   [boolean]

>   -h, --help     Show help                                             [boolean]

Reviewer: why is this part here twice?

Examples:
  mikro-orm schema:update --run  Runs schema synchronization
```

To verify your setup, you can use `mikro-orm debug` command.

> ## Running MikroORM.init() without arguments

Reviewer: IMO this really doesn't belong to quick start, i would leave here just the first sentence and refer to the dedicated CLI guide

> When you have CLI config properly set up, you can omit the `options` parameter when calling `MikroORM.init()` in your app. Configuration is loaded similarly to how it is loaded when using the MikroORM CLI.

Reviewer: ```suggestion
When a CLI config is properly set up, you can omit the `options` parameter when calling `MikroORM.init()` in your app. Configuration is loaded similarly to how it is loaded when using the MikroORM CLI.
```

> The `--config` flag from the command line will be respected also when you run your app (as long as it is part of `process.argv`), not just when you use the CLI.

Reviewer: we should note here that this is now deprecated

```sh

> $ node ./dist/index.js -- --config ./my-orm-config.js

Reviewer: do we really need the `--` in here? i know its not needed for `npx` or all other package managers except `npm`, and i would guess its not needed for running `node` directly either

                                                   [string] [default: "default"]
  -v, --version      Show version number                               [boolean]
  -h, --help         Show help                                         [boolean]

Examples:
  mikro-orm schema:update --run  Runs schema synchronization
```

To verify your setup, you can use `mikro-orm debug` command.

## Running MikroORM.init() without arguments

When you have CLI config properly set up, you can omit the `options` parameter when calling `MikroORM.init()` in your app. The configuration is loaded similarly to how it is loaded when using the MikroORM CLI.

The `--config` flag from the command line will be respected also when you run your app (as long as it is part of `process.argv`), not just when you use the CLI.

```sh
$ node ./dist/index.js -- --config ./my-orm-config.js
```

This might introduce a conflict with other tools like `jest` that also support overriding the config path via `--config` argument. In those cases you can use the `MIKRO_ORM_CONFIG_ARG_NAME` environment variable to change the argument name to something other than `config`:

```sh
$ MIKRO_ORM_CONFIG_ARG_NAME=mikro-orm-config \

>   node ./dist/index.js -- --mikro-orm-config ./my-orm-config.js

Reviewer: same

```

> `jest` does not allow unrecognised parameters, to run tests with a custom configuration you can use this together with `MIKRO_ORM_CLI_CONFIG` environment variable to point to a test config.

> Currently, `process.argv` is analyzed for backwards compatibility. This will be removed in v7. Using `--config` with the MikroORM CLI will still be available.

> By default, TS config files are not considered when running your app. You can use the `alwaysAllowTs` option in your `package.json` file, which will enable checking the TS files even for your actual app, as well as in the MikroORM CLI. This can be handy if you run your app via [Bun](https://bun.sh).

Reviewer: > By default, TS config files are not considered when running your app

this is no longer true, they are considered by default as long as you have `ts-node` installed

Reviewer: The app itself will never register ts-node loader itself, only CLI is doing that, that's because people can leave the ts-node dependency in production builds too for whatever reason, and we don't want to use TS files in there.

Reviewer: Well, it depends. For ESM projects, the loader is needed, and that is what the `mikro-orm-esm` CLI endpoint (or how to call it) is doing. For CJS projects its fine to call `require('ts-node').register()` and that is what the ORM will do for you unless you explicitly disable `useTsNode` in the package.json. This is happening only for the CLI, not for apps. For apps you are responsible to call it yourself, e.g. by replacing `node main.js` with `ts-node main.ts` or similar.

Reviewer: This is the only place where we try to register ts-node:

https://github.com/mikro-orm/mikro-orm/blob/master/packages/cli/src/CLIConfigurator.ts#L51

Reviewer: > Hm... I think things might be made less confusing if ts-node is a required dependency of @mikro-orm/cli

You can use the CLI with compiled files, e.g. in production environments. I would rather not make the CLI dependent on TS.

I know this is rather an edge case, but I believe there are people doing this, e.g. to run compiled migrations in production.

Reviewer: I'd rather leave it like it is now, especially because there are alternatives to ts-node, and in modern runtimes like `bun` you don't even need this since it's supported out of box.

Reviewer: I think the current version is good

    // other credentials and settings
  }),
];
```

Which in turn enables you to run `MikroORM.init()` in your app without arguments (and connect with the user "app"), while in CLI (where you may need higher privileges), you can use
```sh
$ npx mikro-orm --contextName=super
```

You can also export a function, which will be called with a `contextName`, and can give a configuration object for that name, or otherwise return nothing if you wish to error on that name instead. This can be particularly useful in multi-tenant setups.

For example, if you have

```ts
import { defineConfig } from '@mikro-orm/postgresql';

export default (contextName: string) => defineConfig({
  entities: [Author, Book, BookTag],
  dbName: `tenant_${contextName}`,
  user: 'app',
  // other credentials and settings
});
```

then you will need to start your app with the `MIKRO_ORM_CONTEXT_NAME` environment variable set to `example1` to load the database `tenant_example1`, and similarly, when running CLI, you can use

```sh
$ npx mikro-orm --contextName=example1
```

to operate on that particular tenant's database instance. Not specifying either option will point you to the "tenant_default" database.

> You can also combine arrays and factory functions. Array members will be preferred, and any functions in the array will be executed from top to bottom. The first function to return an object will be what ends up being used.

Reviewer: this is nice, quite powerful 👍
